site/en/blog/web-custom-formats-for-the-async-clipboard-api/index.md

```diff
@@ -0,0 +1,98 @@
+---
+layout: 'layouts/blog-post.njk'
+title: Web custom formats for the Async Clipboard API
+subtitle: >
+  Web custom formats let websites read and write arbitrary unsanitized payloads using a standardized
+  format applications can opt in to if they wish to support such payloads.
+authors:
+  - thomassteiner
+date: 2022-07-25
+---
+
+Up until now, the [Async Clipboard API](https://web.dev/async-clipboard/) supported a limited set of MIME types to
+be copied to and pasted from the system clipboard. Namely, these types are `text/plain`,
+`text/html`, and `image/png`, which the browser typically will sanitize, for example, to remove
+embedded `script` elements or `javascript:` links from an HTML string, or to prevent PNG
+[decompression bomb](https://en.wikipedia.org/wiki/Zip_bomb) attacks.
+
+In some cases, though, it can be desirable to support unsanitized content on the clipboard, for
+example, when the application the data is thought for deals itself with the sanitization and where
+it's crucial for the copied data to be identical with the pasted data. For such cases, the Async
+Clipboard API now supports web custom formats that let developers write arbitrary data to the
+clipboard.
+
+## Browser support
+
+The Async Clipboard API per se with image support is supported as of Chromium&nbsp;76. Web
+custom formats for the Async Clipboard API are supported on desktop and on mobile Chromium
+as of version&nbsp;104.
+
+## Writing web custom formats to the clipboard
+
+Writing web custom formats to the clipboard is almost identical to
+[writing sanitized formats](<https://web.dev/async-clipboard/#write()>), except for the requirement
+to prepend the string `"web "` (including the trailing space) to the blob's MIME type.
+
+```js
+// Fetch remote JPEG and GIF images and obtain their blob representations.
+const [jpegBlob, gifBlob] = await Promise.all([
+  fetch('image.jpg').then((response) => response.blob()),
+  fetch('image.gif').then((response) => response.blob()),
+]);
+
+try {
+  // Write the image data to the clipboard, prepending the blobs' actual
+  // types (`"image/jpeg"` and "image/gif") with the string `"web "`, so
+  // they become `"web image/jpeg"` and `"web image/gif"` respectively.
+  const clipboardItem = new ClipboardItem({
+    [`web ${jpegBlob.type}`]: jpegBlob,
+    [`web ${gifBlob.type}`]: gifBlob,
```

Comment:
ClipboardItem constructor does not take an array, but a string. See https://w3c.github.io/clipboard-apis/#clipboarditem 

```suggestion
  const clipboardItem = new ClipboardItem({
    `web ${jpegBlob.type}`: jpegBlob,
    `web ${gifBlob.type}`: gifBlob,
```

Reply:
You may want to update https://web.dev/async-clipboard/#write(), https://web.dev/progressively-enhance-your-pwa/#the-asynchronous-clipboard-api, and https://web.dev/svgcode/#async-clipboard-api as well.

Reply:
I'm wrong! Please dismiss my message.

Reply:
Resolved via https://github.com/GoogleChrome/developer.chrome.com/pull/3249/commits/44f1f49dc94ea1a49c15876ef4bee7d74762b58f.